### PR TITLE
Stream.flat_map/2, Enum.flat_map/2 can use :halt

### PR DESCRIPTION
**Pull request by montague**
Implements #3117

### CODE REVIEW

**Comment by whatyouhide** [comment]
Nice work :+1: Small nitpick: spaces after commas in lists :smiley:.

**Comment by montague** [comment]
good catch--will fix the spaces.

as for using the case statement--i agree, your version is probably more idiomatic elixir. 

will update, thanks for taking a look.

**Comment by whatyouhide** [comment]
@montague no problem, my pleasure!

File: lib/elixir/lib/enum.ex [hunk 1]
```diff
@@ -789,11 +790,19 @@ defmodule Enum do
...
+      ...>    if x > 2, do: :halt, else: [x, x]
+      ...> end)
+      [1, 1, 2, 2]
+
   """
   @spec flat_map(t, (element -> t)) :: list
   def flat_map(collection, fun) do
     reduce(collection, [], fn(entry, acc) ->
-      reduce(fun.(entry), acc, &[&1|&2])
+      case fun.(entry) do
```
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: :+1: for `case` instead of `if`.

Comment (josevalim) [inline lib/elixir/lib/enum.ex]: This code is not correct. Here is an example that will fail:

```
Enum.flat_map([1, 2, 3, 2], fn(x) ->
   if x > 2, do: :halt, else: [x, x]
end)
#=> [1, 1, 2, 2, 2, 2]
```

It should return only `[1, 1, 2, 2]`. To implement this, we would actually need to call `Enumerable.reduce/3`.

Comment (josevalim) [inline lib/elixir/lib/enum.ex]: Wait, if we call `Enumerable.reduce/3` we would actually lose important optimizations. We should not change this I think. Sorry for the false alarm. :(

**Comment by josevalim** [comment]
I feel horrible but to support `:halt` we would need to move to less optimized flat_map which is not worth imo (given one can halt using flat_map_reduce). I am closing this issue.

**Comment by montague** [comment]
ah well. on to the next one :)

**Pull request closed without merging**